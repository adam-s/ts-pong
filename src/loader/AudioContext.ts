// Copyright (c) 2013, Erik Onarheim -- https://github.com/excaliburjs/Excalibur
declare global {
  interface Window {
    webkitAudioContext: typeof AudioContext;
  }
}

export class AudioContextFactory {
  private static _INSTANCE: AudioContext = null;

  public static create(): AudioContext {
    if (!this._INSTANCE) {
      const AudioConstructor = window.AudioContext || window.webkitAudioContext;
      if (AudioConstructor) {
        this._INSTANCE = new AudioConstructor({});
      }
    }

    return this._INSTANCE;
  }
}
