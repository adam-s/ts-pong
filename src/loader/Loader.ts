export interface Loadable<T> {
  data: T;
  load(): Promise<T>;
  isLoaded(): boolean;
}

/**
 * engine.start(loader).then(() => {});
 * ```
 */
export class Loader implements Loadable<Loadable<unknown>[]> {
  private _resourceList: Loadable<unknown>[] = [];
  private _index = 0;

  private _resourceCount = 0;
  private _numLoaded = 0;
  private _progressCounts: { [key: string]: number } = {};
  private _totalCounts: { [key: string]: number } = {};

  constructor(loadables?: Loadable<unknown>[]) {
    if (loadables) {
      this.addResources(loadables);
    }
  }

  public addResource(loadable: Loadable<unknown>) {
    const key = this._index++;
    this._resourceList.push(loadable);
    this._progressCounts[key] = 0;
    this._totalCounts[key] = 1;
    this._resourceCount++;
  }

  public addResources(loadables: Loadable<unknown>[]) {
    let i = 0;
    const len = loadables.length;

    for (i; i < len; i++) {
      this.addResource(loadables[i]);
    }
  }

  public isLoaded() {
    return this._numLoaded === this._resourceCount;
  }

  data: Loadable<unknown>[];

  public async load(): Promise<Loadable<unknown>[]> {
    await Promise.all(
      this._resourceList.map(async (r) => {
        try {
          return await r.load();
        } catch (error) {
          console.log('Loading error: ', error);
        } finally {
          // capture progress
          this._numLoaded++;
        }
      })
    );
    return (this.data = this._resourceList);
  }

  public markResourceComplete(): void {
    this._numLoaded++;
  }
}
