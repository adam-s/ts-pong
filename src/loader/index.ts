export * from './Loader';
export * from './Resource';
export * from './ImageSource';
export * from './AudioSource';
export * from './AudioContext';
export * from './WebAudioInstance';
