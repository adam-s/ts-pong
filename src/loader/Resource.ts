import { EventDispatcher } from '../utils';
import { Loadable } from './Loader';

export class Resource<T> implements Loadable<T> {
  public data: T = null;
  public events: EventDispatcher = new EventDispatcher();

  constructor(
    public path: string,
    public responseType:
      | ''
      | 'arraybuffer'
      | 'blob'
      | 'document'
      | 'json'
      | 'text',
    public bustCache: boolean = true
  ) {}

  // This is fun. Use on and emit to delegate to the eventing. Need to bind the events instance to the on method though.
  get on() {
    return this.events.on.bind(this.events);
  }

  get emit() {
    return this.events.emit.bind(this.events);
  }

  public isLoaded(): boolean {
    return this.data !== null;
  }

  private _cacheBust(uri: string): string {
    const query = /\?\w*=\w*/;
    if (query.test(uri)) {
      uri += '&__=' + Date.now();
    } else {
      uri += '?__=' + Date.now();
    }
    return uri;
  }

  /**
   * Begin loading the resource and returns a promise to be resolved on completion
   */
  public load(): Promise<T> {
    return new Promise((resolve, reject) => {
      // Exit early if we already have data
      if (this.data !== null) {
        this.emit('complete', this.data as unknown);
        resolve(this.data);
        return;
      }

      const request = new XMLHttpRequest();
      request.open(
        'GET',
        this.bustCache ? this._cacheBust(this.path) : this.path,
        true
      );
      request.responseType = this.responseType;
      request.addEventListener('loadstart', (e) =>
        this.emit('loadstart', e as unknown)
      );
      request.addEventListener('progress', (e) =>
        this.emit('progress', e as unknown)
      );
      request.addEventListener('error', (e) =>
        this.emit('error', e as unknown)
      );
      request.addEventListener('load', (e) => this.emit('load', e as unknown));
      request.addEventListener('load', () => {
        // XHR on file:// success status is 0, such as with PhantomJS
        if (request.status !== 0 && request.status !== 200) {
          this.emit('error', request.response);
          reject(new Error(request.statusText));
          return;
        }

        this.data = request.response;
        this.emit('complete', this.data as unknown);
        resolve(this.data);
      });
      request.send();
    });
  }
}
