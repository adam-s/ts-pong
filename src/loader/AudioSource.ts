// Copyright (c) 2013, Erik Onarheim -- https://github.com/excaliburjs/Excalibur
import { canPlayFile } from '../utils';
import { AudioContextFactory } from './AudioContext';
import { Loadable } from './Loader';
import { Resource } from './Resource';
import { WebAudioInstance } from './WebAudioInstance';

export interface Audio {
  /**
   * Whether the audio should loop (repeat forever)
   */
  loop: boolean;
  /**
   * The volume (between 0 and 1)
   */
  volume: number;

  /**
   * Whether or not any audio is playing
   */
  isPlaying(): boolean;

  /**
   * Will play the sound or resume if paused
   */
  play(): Promise<boolean>;

  /**
   * Pause the sound
   */
  pause(): void;

  /**
   * Stop playing the sound and reset
   */
  stop(): void;
}

export class AudioSource implements Audio, Loadable<AudioBuffer> {
  public data: AudioBuffer;
  private _resource: Resource<ArrayBuffer>;
  /**
   * Indicates whether the clip should loop when complete
   * @param value  Set the looping flag
   */
  public set loop(value: boolean) {
    this._loop = value;
  }
  public get loop(): boolean {
    return this._loop;
  }

  public set volume(value: number) {
    this._volume = value;
  }
  public get volume(): number {
    return this._volume;
  }

  public get duration(): number | undefined {
    return this._duration;
  }

  /**
   * Return array of Current AudioInstances playing or being paused
   */
  public get instances(): Audio[] {
    return this._tracks;
  }

  public get path() {
    return this._resource.path;
  }

  public set path(val: string) {
    this._resource.path = val;
  }

  private _loop = false;
  private _volume = 1;
  private _duration: number | undefined = undefined;
  private _isStopped = false;
  private _isPaused = false;
  private _tracks: Audio[] = [];
  private _audioContext = AudioContextFactory.create();

  /**
   * @param paths A list of audio sources (clip.wav, clip.mp3, clip.ogg) for this audio clip. This is done for browser compatibility.
   */
  constructor(...paths: string[]) {
    this._resource = new Resource('', 'arraybuffer');
    for (const path of paths) {
      if (canPlayFile(path)) {
        this.path = path;
        break;
      }
    }
  }

  public isLoaded() {
    return !!this.data;
  }

  public async load(): Promise<AudioBuffer> {
    if (this.data) {
      return this.data;
    }
    const arraybuffer = await this._resource.load();
    const audiobuffer = await this.decodeAudio(arraybuffer.slice(0));
    this._duration =
      typeof audiobuffer === 'object' ? audiobuffer.duration : undefined;
    return (this.data = audiobuffer);
  }

  public async decodeAudio(data: ArrayBuffer): Promise<AudioBuffer> {
    try {
      return await this._audioContext.decodeAudioData(data.slice(0));
    } catch (e) {
      console.error(`Unable to decode this browser may not fully support this format,\
 or the file may be corrupt, if this is an mp3 try removing id3\
 tags and album art from the file.`);
      return await Promise.reject();
    }
  }

  /**
   * Returns how many instances of the sound are currently playing
   */
  public instanceCount(): number {
    return this._tracks.length;
  }

  /**
   * Whether or not the sound is playing right now
   */
  public isPlaying(): boolean {
    return this._tracks.some((t) => t.isPlaying());
  }

  /**
   * Play the sound, returns a promise that resolves when the sound is done playing
   * An optional volume argument can be passed in to play the sound. Max volume is 1.0
   */
  public play(volume?: number): Promise<boolean> {
    if (!this.isLoaded()) {
      console.warn(
        'Cannot start playing. Resource',
        this.path,
        'is not loaded yet'
      );

      return Promise.resolve(true);
    }

    if (this._isStopped) {
      console.warn('Cannot start playing. Engine is in a stopped state.');
      return Promise.resolve(false);
    }

    this.volume = volume || this.volume;

    if (this._isPaused) {
      return this._resumePlayback();
    } else {
      return this._startPlayback();
    }
  }

  /**
   * Stop the sound, and do not rewind
   */
  public pause() {
    if (!this.isPlaying()) {
      return;
    }

    for (const track of this._tracks) {
      track.pause();
    }

    this._isPaused = true;
  }

  /**
   * Stop the sound if it is currently playing and rewind the track. If the sound is not playing, rewinds the track.
   */
  public stop() {
    for (const track of this._tracks) {
      track.stop();
    }

    this._isPaused = false;
    this._tracks.length = 0;
  }

  /**
   * Get Id of provided AudioInstance in current trackList
   * @param track [[Audio]] which Id is to be given
   */
  public getTrackId(track: Audio): number {
    return this._tracks.indexOf(track);
  }

  private async _resumePlayback(): Promise<boolean> {
    if (this._isPaused) {
      const resumed: Promise<boolean>[] = [];
      // ensure we resume *current* tracks (if paused)
      for (const track of this._tracks) {
        resumed.push(track.play());
      }
      this._isPaused = false;
      // resolve when resumed tracks are done
      await Promise.all(resumed);
    }
    return true;
  }

  /**
   * Starts playback, returns a promise that resolves when playback is complete
   */
  private async _startPlayback(): Promise<boolean> {
    const track = await this._getTrackInstance(this.data);

    const complete = await track.play();

    // when done, remove track
    this._tracks.splice(this.getTrackId(track), 1);

    return complete;
  }

  private _getTrackInstance(data: AudioBuffer): WebAudioInstance {
    const newTrack = new WebAudioInstance(data);

    newTrack.loop = this.loop;
    newTrack.volume = this.volume;
    newTrack.duration = this.duration;

    this._tracks.push(newTrack);

    return newTrack;
  }
}
