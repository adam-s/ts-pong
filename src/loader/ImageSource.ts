import { Loadable } from './Loader';
import { Resource } from './Resource';

export class ImageSource implements Loadable<HTMLImageElement> {
  private _resource: Resource<Blob>;

  public get width() {
    return this.image.naturalWidth;
  }

  public get height() {
    return this.image.naturalHeight;
  }

  public isLoaded(): boolean {
    return !!this.data.src;
  }

  public data: HTMLImageElement = new Image();
  public get image(): HTMLImageElement {
    return this.data;
  }

  public ready: Promise<HTMLImageElement>;
  private _loadedResolve: (
    value?: HTMLImageElement | PromiseLike<HTMLImageElement>
  ) => void;

  constructor(public readonly path: string, bustCache = false) {
    this._resource = new Resource(path, 'blob', bustCache);
    if (path.endsWith('.svg') || path.endsWith('.gif')) {
      console.warn(
        `Image type is not fully supported, you may have mixed results ${path}. Fully supported: jpg, bmp, and png`
      );
    }
    this.ready = new Promise<HTMLImageElement>((resolve) => {
      this._loadedResolve = resolve;
    });
  }

  async load(): Promise<HTMLImageElement> {
    if (this.isLoaded()) {
      return this.data;
    }

    try {
      // Load base64 or blob if needed
      let url: string;
      if (!this.path.includes('data:image/')) {
        const blob = await this._resource.load();
        url = URL.createObjectURL(blob);
      } else {
        url = this.path;
      }
      // Decode the image
      const image = new Image();
      image.src = url;
      await image.decode();

      // Set results
      this.data = image;
    } catch (error) {
      throw `Error loading ImageSource from path '${this.path}' with error [${error.message}]`;
    }
    // todo emit complete
    this._loadedResolve(this.data);
    return this.data;
  }

  unload(): void {
    this.data = new Image();
  }
}
