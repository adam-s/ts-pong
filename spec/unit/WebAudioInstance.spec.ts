import './AudioContext.mock';
import { WebAudioInstance } from '../../src/loader/WebAudioInstance';

describe('WebAudioInstance', () => {
  let implementation: WebAudioInstance;
  let createNewBufferSourceSpy: jest.SpyInstance;
  let rememberStartTimeSpy: jest.SpyInstance;
  let volumeNodeConnectSpy: jest.SpyInstance;
  let startSpy: jest.SpyInstance;
  let wireUpOnEndedSpy: jest.SpyInstance;

  beforeEach(() => {
    implementation = new WebAudioInstance(
      new AudioBuffer({ length: 6, sampleRate: 1 })
    );

    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    createNewBufferSourceSpy = jest.spyOn(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      implementation as any,
      '_createNewBufferSource'
    );
    rememberStartTimeSpy = jest.spyOn(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      implementation as any,
      '_rememberStartTime'
    );
    volumeNodeConnectSpy = jest.spyOn(implementation['_volumeNode'], 'connect');
    startSpy = jest.spyOn(implementation['_instance'], 'start');
    wireUpOnEndedSpy = jest.spyOn(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      implementation as any,
      '_wireUpOnEnded'
    );
    jest.clearAllMocks();
  });

  it('should exist', () => {
    expect(WebAudioInstance).toBeDefined();
  });

  it('can be created', () => {
    expect(implementation).toBeDefined();
    expect(implementation).toMatchSnapshot();
  });

  it('loops', () => {
    const wireUpSpy = jest.spyOn(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      implementation as any,
      '_wireUpOnEnded'
    );
    implementation.loop = true;
    expect(wireUpSpy).toBeCalledTimes(1);
  });

  it('volume is clamped to 0', () => {
    implementation.volume = 0.5;
    expect(implementation.volume).toBe(0.5);
    implementation.volume = -1;
    expect(implementation.volume).toBe(0);
  });

  it('volume is clamped to 1', () => {
    implementation.volume = 0.5;
    expect(implementation.volume).toBe(0.5);
    implementation.volume = 2;
    expect(implementation.volume).toBe(1);
  });

  it('calls setTargetAtTime if playing and volume is set', () => {
    const setTargetAtTimeSpy = jest.spyOn(
      implementation['_volumeNode'].gain,
      'setTargetAtTime'
    );
    implementation['_isPlaying'] = true;
    implementation.volume = 0.5;
    expect(setTargetAtTimeSpy).toBeCalledWith(0.5, 1, 0.1);
  });

  it('gets playback rate', () => {
    expect(implementation['_playbackRate']).toBe(1);
  });

  it('plays if paused', () => {
    const playStarted = jest.fn();
    implementation['_isPaused'] = true;
    implementation.play(playStarted);
    expect(playStarted).toBeCalled();
  });

  it('plays if not playing', () => {
    const playStarted = jest.fn();
    implementation['_isPlaying'] = false;
    implementation.play(playStarted);
    expect(playStarted).toBeCalled();
  });

  it('pauses', () => {
    const stopSpy = jest.spyOn(implementation['_instance'], 'stop');
    const setPauseOffsetSpy = jest.spyOn(
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      implementation as any,
      '_setPauseOffset'
    );
    implementation['_isPlaying'] = true;
    implementation.pause();
    expect(stopSpy).toBeCalledWith(0);
    expect(setPauseOffsetSpy).toBeCalled();
    expect(implementation['_isPaused']).toBe(true);
    expect(implementation['_isPlaying']).toBe(false);
  });

  it('stops', () => {
    const stopSpy = jest.spyOn(implementation['_instance'], 'stop');
    implementation['_isPlaying'] = true;
    implementation.stop();
    expect(stopSpy).toBeCalledWith(0);
    expect(implementation['_isPaused']).toBe(false);
    expect(implementation['_isPlaying']).toBe(false);
    expect(implementation['_currentOffset']).toBe(0);
  });

  it('starts playback', () => {
    implementation['_isPlaying'] = false;
    implementation['_isPaused'] = true;
    implementation['_currentOffset'];
    implementation['_startPlayBack']();
    expect(implementation['_isPlaying']).toBe(true);
    expect(implementation['_isPaused']).toBe(false);
    expect(rememberStartTimeSpy).toBeCalled();
    expect(volumeNodeConnectSpy).toBeCalledWith(
      implementation['_audioContext'].destination
    );
    expect(startSpy).toBeCalledWith(0, 0);
    expect(implementation['_currentOffset']).toBe(0);
    expect(wireUpOnEndedSpy).toBeCalled();
    delete implementation['_instance'];
    expect(createNewBufferSourceSpy).not.toBeCalled();
    implementation['_startPlayBack']();
    expect(createNewBufferSourceSpy).toBeCalled();
  });

  it('resumes play back', () => {
    implementation['_isPaused'] = true;
    implementation['_isPlaying'] = false;
    implementation['_currentOffset'] = 50;
    implementation['_resumePlayBack']();
    expect(implementation['_isPaused']).toBe(false);
    expect(implementation['_isPlaying']).toBe(true);
    expect(createNewBufferSourceSpy).toBeCalled();
    expect(rememberStartTimeSpy).toBeCalledWith(-50000);
    expect(startSpy).toBeCalledWith(0, 50);
    expect(wireUpOnEndedSpy).toBeCalled();
  });

  it('wires up on ended', () => {
    implementation['loop'] = false;
    implementation['_instance'].onended = null;
    implementation['_wireUpOnEnded']();
    expect(typeof implementation['_instance'].onended).toBe('function');
  });

  it("remember's start time", () => {
    jest.useFakeTimers().setSystemTime(new Date('2020-01-01').getTime());
    implementation['_rememberStartTime'](1000);
    expect(implementation['_startTime']).toBe(1577836801000);
  });

  it('sets pause offset', () => {
    expect(implementation['_currentOffset']).toBe(0);
    jest.useFakeTimers().setSystemTime(new Date('2020-01-01').getTime());
    implementation['_startTime'] = 1;
    implementation['_setPauseOffset']();
    expect(implementation['_currentOffset']).toBe(1577836799.999);
  });

  it('creates a new buffer source', () => {
    const mockSetValueAtTime = jest.fn();
    const mockConnect = jest.fn();
    const createBufferSourceSpy = jest
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      .spyOn(implementation['_audioContext'] as any, 'createBufferSource')
      .mockReturnValue({
        buffer: null,
        loop: null,
        playbackRate: {
          setValueAtTime: mockSetValueAtTime,
        },
        connect: mockConnect,
      });
    implementation['_createNewBufferSource']();
    expect(createBufferSourceSpy).toBeCalled();
    expect(implementation['_instance'].buffer).toBe(implementation['_src']);
    expect(implementation['_instance'].loop).toBe(implementation['loop']);
    expect(mockSetValueAtTime).toBeCalledWith(1.0, 0);
    expect(mockConnect).toBeCalledWith(implementation['_volumeNode']);
  });
});
