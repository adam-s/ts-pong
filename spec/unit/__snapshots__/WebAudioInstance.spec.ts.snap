// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebAudioInstance can be created 1`] = `
WebAudioInstance {
  "_audioContext": Object {
    "createBufferSource": [Function],
    "createGain": [MockFunction],
    "currentTime": 1,
    "decodeAudioData": [MockFunction],
    "destination": "DESTINATION_NODE",
  },
  "_createNewBufferSource": [MockFunction],
  "_currentOffset": 0,
  "_duration": undefined,
  "_instance": Object {
    "buffer": Object {
      "duration": 5000,
      "gain": 1,
      "length": 6,
      "numberOfChannels": undefined,
      "sampleRate": 1,
    },
    "connect": [MockFunction],
    "onended": [MockFunction],
    "playbackRate": Object {
      "setValueAtTime": [MockFunction],
      "value": 1,
    },
    "start": [MockFunction],
    "stop": [MockFunction],
  },
  "_isPaused": false,
  "_isPlaying": false,
  "_loop": false,
  "_rememberStartTime": [MockFunction],
  "_src": Object {
    "duration": 5000,
    "gain": 1,
    "length": 6,
    "numberOfChannels": undefined,
    "sampleRate": 1,
  },
  "_volume": 1,
  "_volumeNode": Object {
    "connect": [MockFunction],
    "gain": Object {
      "setTargetAtTime": [MockFunction],
      "value": 1,
    },
  },
  "_wireUpOnEnded": [MockFunction],
}
`;
